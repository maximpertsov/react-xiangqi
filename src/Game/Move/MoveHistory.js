import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import Move from './Move';

// TODO: set max-height by percentage?
const Wrapper = styled.div`
  display: grid;
  grid-template-columns: 50% auto;
  grid-template-rows: repeat(auto-fill, 50px);
  outline: thin solid #999;
  height: 55%;
  overflow: auto;
`;

const MoveHistory = ({ moves, selectedMoveIdx, handleMoveSelect }) => {
  const scrollTarget = (<div ref={(el) => { this.el = el; }} />);
  const moveComponents = moves
    .map((m, i) => (
      <Move
        key={i}
        idx={i}
        handleMoveSelect={handleMoveSelect}
        fromPos={m.fromPos}
        toPos={m.toPos}
        piece={m.piece}
        selected={selectedMoveIdx === i}
      />
    ));

  return (
    <Wrapper>
      {moveComponents}
      {scrollTarget}
    </Wrapper>
  );
};

Move.propTypes = {
  // TODO: add move proptype
  moves: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  selectedMoveIdx: PropTypes.number.isRequired,
  handleMoveSelect: PropTypes.func.isRequired,
};

export default MoveHistory;
