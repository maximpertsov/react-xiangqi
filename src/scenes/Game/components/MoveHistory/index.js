import React from 'react';
import { useSelector } from 'react-redux';
import styled from '@emotion/styled';
import { Segment } from 'semantic-ui-react';
import { tail, chunk, isEqual } from 'lodash';

import { MediaQuery, WidthSize } from 'commonStyles';
import Move from './components/Move';
import FullMove from './components/FullMove';

const repeatItemCSS = times =>
  Array(times)
    .fill('1fr')
    .join(' ');

const Wrapper = styled.div`
  display: grid;
  ${MediaQuery.TINY} {
    grid-template-columns: ${repeatItemCSS(2)};
    font-size: x-small;
    width: ${WidthSize.TINY};
  }
  ${MediaQuery.SMALL} {
    grid-template-columns: ${repeatItemCSS(2)};
    font-size: x-small;
    width: ${WidthSize.SMALL};
  }
  ${MediaQuery.MEDIUM} {
    grid-template-columns: ${repeatItemCSS(2)};
    font-size: small;
    width: ${WidthSize.MEDIUM};
  }
  ${MediaQuery.LARGE} {
    grid-template-columns: ${repeatItemCSS(3)};
    font-size: small;
    width: ${WidthSize.LARGE};
  }
  grid-template-rows: repeat(auto-fill, 1fr);
  font-size: small;
  max-height: 10vh;
  overflow: auto;
`;

const MoveHistory = () => {
  const positions = useSelector(state => state.positions, isEqual);

  const positionComponents = positions.map((position, index) => (
    <Move
      key={index}
      moveId={position.id}
      fan={position.fan}
      fen={position.fen}
    />
  ));
  const fullMoves = chunk(tail(positionComponents), 2).map(
    ([redMove, blackMove], index) => (
      <FullMove key={index} ordering={index + 1}>
        {redMove}
        {blackMove}
      </FullMove>
    ),
  );

  return (
    <Segment clearing tertiary>
      <Wrapper className="MoveHistory">{fullMoves}</Wrapper>
    </Segment>
  );
};

export default MoveHistory;
