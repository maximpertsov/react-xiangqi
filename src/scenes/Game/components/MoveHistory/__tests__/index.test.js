import React from 'react';
import MoveHistory from 'scenes/Game/components/MoveHistory';

describe('MoveHistory', () => {
  let store;
  let wrapper;
  let spys = {};

  beforeEach(() => {
    setupReduxSpys(spys, store);
    wrapper = shallowWrappedComponent(<MoveHistory />, store);
  });

  afterEach(() => {
    store.clearActions();
    restoreSpys(spys);
  });

  describe('render', () => {
    beforeAll(() => {
      store = mockStore({
        moves: [{ fen: 'FEN', uci: 'a1a2' }],
      });
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
