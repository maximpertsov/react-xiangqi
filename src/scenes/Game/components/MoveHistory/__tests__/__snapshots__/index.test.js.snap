// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveHistory render snapshot 1`] = `
<Segment
  clearing={true}
  tertiary={true}
>
  <Styled(div)
    className="MoveHistory"
  />
</Segment>
`;
